File: src/games/blockblast/sound.js
// Block Blast sound — tiny WebAudio synthesizer, same pattern as the Tetris
// sfx module. Zero asset dependencies; every effect is a short oscillator
// envelope. Nothing plays until the first call creates the AudioContext.

let ctx = null;
let masterGain = null;
let enabled = true;

function getCtx() {
  if (ctx) return ctx;
  try {
    const AC = window.AudioContext || window.webkitAudioContext;
    if (!AC) return null;
    ctx = new AC();
    masterGain = ctx.createGain();
    masterGain.gain.value = 0.22;
    masterGain.connect(ctx.destination);
  } catch (_) {
    return null;
  }
  return ctx;
}

function tone(freq, durationMs, type = 'square', { attack = 0.005, release = 0.08, gain = 0.7 } = {}) {
  if (!enabled) return;
  const ac = getCtx();
  if (!ac) return;
  if (ac.state === 'suspended') { void ac.resume().catch(() => {}); }
  const osc = ac.createOscillator();
  const env = ac.createGain();
  osc.type = type;
  osc.frequency.value = freq;
  const now = ac.currentTime;
  env.gain.setValueAtTime(0, now);
  env.gain.linearRampToValueAtTime(gain, now + attack);
  env.gain.exponentialRampToValueAtTime(0.0001, now + durationMs / 1000 + release);
  osc.connect(env);
  env.connect(masterGain);
  osc.start(now);
  osc.stop(now + durationMs / 1000 + release + 0.02);
}

function sweep(from, to, durationMs, type = 'triangle', { gain = 0.5 } = {}) {
  if (!enabled) return;
  const ac = getCtx();
  if (!ac) return;
  if (ac.state === 'suspended') { void ac.resume().catch(() => {}); }
  const osc = ac.createOscillator();
  const env = ac.createGain();
  osc.type = type;
  const now = ac.currentTime;
  osc.frequency.setValueAtTime(from, now);
  osc.frequency.exponentialRampToValueAtTime(to, now + durationMs / 1000);
  env.gain.setValueAtTime(0, now);
  env.gain.linearRampToValueAtTime(gain, now + 0.01);
  env.gain.exponentialRampToValueAtTime(0.0001, now + durationMs / 1000 + 0.05);
  osc.connect(env);
  env.connect(masterGain);
  osc.start(now);
  osc.stop(now + durationMs / 1000 + 0.1);
}

export const sfx = {
  pickUp()   { tone(520, 40, 'triangle', { gain: 0.3 }); },
  place()    { tone(200, 50, 'sawtooth', { gain: 0.35 }); },
  invalid()  { tone(110, 80, 'square', { gain: 0.25 }); },
  clear1()   { sweep(440, 660, 140, 'triangle', { gain: 0.45 }); },
  clear2()   {
    sweep(440, 880, 160, 'triangle', { gain: 0.5 });
    setTimeout(() => sweep(660, 1100, 180, 'triangle', { gain: 0.45 }), 90);
  },
  clearBig() {
    sweep(330, 660, 90, 'sawtooth', { gain: 0.45 });
    setTimeout(() => sweep(660, 1320, 140, 'sawtooth', { gain: 0.5 }), 70);
    setTimeout(() => sweep(1320, 1760, 180, 'sawtooth', { gain: 0.55 }), 180);
  },
  combo()    {
    tone(660, 60, 'triangle', { gain: 0.4 });
    setTimeout(() => tone(990, 60, 'triangle', { gain: 0.45 }), 50);
  },
  levelUp()  {
    tone(440, 80, 'triangle', { gain: 0.4 });
    setTimeout(() => tone(660, 80, 'triangle', { gain: 0.42 }), 80);
    setTimeout(() => tone(880, 140, 'triangle', { gain: 0.48 }), 160);
  },
  gameOver() {
    sweep(440, 110, 600, 'sawtooth', { gain: 0.5 });
  },
  start()    {
    tone(440, 60, 'triangle', { gain: 0.35 });
    setTimeout(() => tone(660, 60, 'triangle', { gain: 0.35 }), 70);
    setTimeout(() => tone(880, 120, 'triangle', { gain: 0.4 }), 140);
  },
};

export function setSoundEnabled(v) { enabled = !!v; }
export function isSoundEnabled() { return enabled; }

File: lib/lib/games/blockblast/sound.dart
// Port of src/games/blockblast/sound.js — Block Blast SFX helper.
//
// The JS version is a tiny WebAudio synthesizer: every effect is a short
// oscillator envelope, no asset files. There's no equivalent low-level audio
// graph API shipped with Flutter, and the current pubspec doesn't depend on
// a synth/audio package, so this is a NO-OP STUB that preserves the public
// API (`sfx.place()`, `sfx.clear1()`, etc. and `setSoundEnabled`).
//
// TODO to make these actually play:
//   1. Add `audioplayers: ^6.0.0` (or `just_audio`) to pubspec.yaml.
//   2. Ship short WAV/OGG samples in `assets/sfx/blockblast/`.
//   3. Replace the `_noop` bodies below with `AudioPlayer().play(...)` calls.
// Until then the game is silent; haptics (HapticFeedback via haptics.dart)
// still fire so placement/clear feedback is not completely missing.

bool _enabled = true;

void setSoundEnabled(bool v) {
  _enabled = v;
}

bool isSoundEnabled() => _enabled;

void _noop() {
  // If sound were wired up this is where the check would branch. Left as a
  // single function so future `if (!_enabled) return;` guards live in one
  // place when we swap the stub for real playback.
}

/// Block Blast SFX bank. Every method mirrors a callable on the JS `sfx`
/// object in `sound.js`. All are no-ops for now — see file header for how
/// to hook up real audio.
class _BlockBlastSfx {
  const _BlockBlastSfx();

  void pickUp() => _noop();   // tone 520Hz, 40ms, triangle
  void place() => _noop();    // tone 200Hz, 50ms, sawtooth
  void invalid() => _noop();  // tone 110Hz, 80ms, square
  void clear1() => _noop();   // sweep 440 → 660, 140ms, triangle
  void clear2() => _noop();   // two-step sweep 440 → 880 then 660 → 1100
  void clearBig() => _noop(); // three-step sweep climax for >= 3 lines
  void combo() => _noop();    // tone 660 then 990, triangle
  void levelUp() => _noop();  // 440 → 660 → 880 triangle arpeggio
  void gameOver() => _noop(); // sweep 440 → 110, 600ms, sawtooth
  void start() => _noop();    // 440 → 660 → 880 triangle arpeggio
}

/// Singleton instance the widget reaches for — mirrors `import { sfx }` in
/// the JS source. Keeping it as a const instance rather than top-level
/// functions matches the JS namespacing so porting is 1:1.
const _BlockBlastSfx sfx = _BlockBlastSfx();
